compiler/rustc_hir/src/def.rs

```diff
@@ -314,68 +314,75 @@ pub enum Res<Id = hir::HirId> {
     /// **Belongs to the type namespace.**
     PrimTy(hir::PrimTy),
 
-    /// The `Self` type, optionally with the [`DefId`] of the trait it belongs to and
-    /// optionally with the [`DefId`] of the item introducing the `Self` type alias.
+    /// The `Self` type, as used within a trait.
     ///
     /// **Belongs to the type namespace.**
     ///
-    /// Examples:
+    /// Example:
     /// ```
-    /// struct Bar(Box<Self>);
-    /// // `Res::SelfTy { trait_: None, alias_of: Some(Bar) }`
-    ///
     /// trait Foo {
-    ///     fn foo() -> Box<Self>;
-    ///     // `Res::SelfTy { trait_: Some(Foo), alias_of: None }`
+    ///     fn foo() -> Box<Self>; // `Res::SelfTyTrait { trait_: Foo }`
     /// }
+    /// ```
+    SelfTyTrait {
```

Comment:
```suggestion
    SelfTyParam {
```
For consistency with `DefKind::TyParam`, of which this variant is an equivalent.

Reply:
I don't understand this suggestion. The comment on `DefKind::TyParam` says:
```rust
    /// Type parameter: the `T` in `struct Vec<T> { ... }`
```
`Res::SelfTyTrait` is for `Self` within a trait. Isn't that totally different?

Reply:
No, semantically `Self` is the first type parameter in any trait.
The only difference is that it's introduced implicitly rather than through explicit `<T1, T2, T3>` syntax (and correspondingly cannot be used inside `<>`s when generic arguments are provided).

Reply:
This is what the comment in https://github.com/rust-lang/rust/pull/101887#discussion_r972938005 is about.